Adding Image captioning to Pythia

vedanuj opened the pull request:
Adding image captioning task and Bottom Up Top Down attention model [link](https://arxiv.org/abs/1707.07998) into Pythia Framework. Model trained with the `butd.yml` config reaches close to the numbers reported in the paper. Also adds beam search decoding implementation. However this is not integrated to the standard flow yet. We can add something similar to `evalai` report generation for vqa datasets[needs discussion]. 

Currently we have added `bleu4` score from `nltk` library for `caption_accuracy` calculation.

apsdehal commented on pythia/modules/decoders.py:
```diff
@@ -47,3 +49,78 @@ def forward(self, encoder_output, batch):
+        features_dim=2048,
+        dropout=0.5,
+    ):
+        super(CaptioningDecoder, self).__init__()
```
NIT: Be python 3 advocate and change this to `super().__init__()` maybe.

apsdehal commented on pythia/modules/decoders.py:
```diff
@@ -47,3 +49,78 @@ def forward(self, encoder_output, batch):
+        self.init_weights()
+     
+    def init_weights(self):
+        self.fc.bias.data.fill_(0)
```
These should also be configurable from the configuration.

apsdehal commented on pythia/modules/decoders.py:
```diff
@@ -47,3 +49,78 @@ def forward(self, encoder_output, batch):
+        self.fc.weight.data.uniform_(-0.1, 0.1)
+
+    def init_hidden_state(self, batch_size):
+        h = torch.zeros(batch_size, self.decoder_dim).to("cuda")
```
As mentioned above, please use `new` creation ops or use this and move to proper device.

apsdehal commented on pythia/modules/decoders.py:
```diff
@@ -47,3 +49,78 @@ def forward(self, encoder_output, batch):
+        c = torch.zeros(batch_size, self.decoder_dim).to("cuda")
+        return h, c
+
+    def forward(self, image_features, caption_embeddings, state=None):
```
As discussed above, top-down attention class should be modified for this while being also usable for VQA and others. Decoder should technically be only generating the required text.

vedanuj replied:
Do you mean keeping only the `Language LSTM` part as the decoder?

apsdehal commented on pythia/modules/losses.py:
```diff
@@ -200,6 +201,32 @@ def forward(self, sample_list, model_output):
         return loss * targets.size(1)
 
+@registry.register_loss("cross_entropy")
+class CrossEntropyLoss(nn.Module):
```
Let's maybe call it something else. This is not a generic `CrossEntropyLoss` and is too specific to captioning and in general generation.

vedanuj replied:
Will rename to `CaptionCrossEntropyLoss`

apsdehal commented on pythia/modules/metrics.py:
```diff
@@ -49,6 +49,7 @@ def calculate(self, sample_list, model_output):
 import torch
 
 from pythia.common.registry import registry
+from nltk.translate.bleu_score import corpus_bleu
```
Let's not import it here. Import it only when your metric is initialized. This way this doesn't become a hard dependency.

apsdehal commented on pythia/modules/metrics.py:
```diff
@@ -198,6 +199,86 @@ def calculate(self, sample_list, model_output, *args, **kwargs):
 
 
+@registry.register_metric("caption_accuracy")
+class CaptionAccuracy(BaseMetric):
```
Is there some way it can be more generic and call handle any text strings and probabilities. I believe it requires minor changes. You will probably have to document in more detail.

Important: Please divide metrics into separate classes. This should be called `Bleu4` or `Blue4ForCaption` and you should implement other metrics separately so they are easily reusable.

apsdehal commented on pythia/modules/metrics.py:
```diff
@@ -198,6 +199,86 @@ def calculate(self, sample_list, model_output, *args, **kwargs):
+            img_caps = targets[j].tolist()
+            img_captions = list(
+                map(
+                    lambda c: [
```
Let's stay away from lambda functions. They are hard to test and can easily create issues. Can you move this to separate function and follow comments about readability as above?

apsdehal commented on pythia/modules/metrics.py:
```diff
@@ -198,6 +199,86 @@ def calculate(self, sample_list, model_output, *args, **kwargs):
+
+        # Hypotheses
+        scores = model_output["scores"]
+        _, scores = torch.max(scores, dim=2)
```
NIT: Might be better to use `dim=-1` to be more generic.

apsdehal commented on pythia/modules/metrics.py:
```diff
@@ -198,6 +199,86 @@ def calculate(self, sample_list, model_output, *args, **kwargs):
+        scores = scores.tolist()
+        predictions = list()
+        for j, p in enumerate(scores):
+            for idx, v in enumerate(scores[j]):
```
NIT: Better variables names for readability.

vedanuj replied:
I will move this logic to a caption processor.

apsdehal commented on pythia/modules/metrics.py:
```diff
@@ -198,6 +199,86 @@ def calculate(self, sample_list, model_output, *args, **kwargs):
+    **Key:** ``caption_accuracy``
+    """
+
+    def __init__(self):
```
Import nltk here.

apsdehal commented on pythia/tasks/captioning/coco/builder.py:
```diff
@@ -0,0 +1,84 @@
+
+
+@registry.register_builder("coco")
+class COCOBuilder(BaseDatasetBuilder):
```
Isn't this very similar to VQA2Builder? Why not inherit that class?

apsdehal commented on pythia/tasks/captioning/coco/dataset.py:
```diff
@@ -0,0 +1,152 @@
+from pythia.utils.general import get_pythia_root
+
+
+class COCODataset(BaseDataset):
```
It probably will be better to inherit VQA2Dataset class here and use most of the similar code from it.

apsdehal commented on pythia/tasks/captioning/coco/dataset.py:
```diff
@@ -0,0 +1,152 @@
+    def __len__(self):
+        return len(self.imdb)
+
+    def try_fast_read(self):
```
Please use VQA2Dataset class so we can reuse most of the code.

apsdehal commented on pythia/tasks/captioning/coco/dataset.py:
```diff
@@ -0,0 +1,152 @@
+                if el == vocab.EOS_INDEX:
+                    captions[idx] = captions[idx][:idx_]
+                    break
+            answer = [
```
Readability.

vedanuj commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+
+        targets = sample_list.answers[:, 0]
+        sample_list.add_field("targets", targets)
+        sample_list.add_field("captions", captions)
```
`sample_list` requires a `targets` field otherwise it doesn't calculate loss. The `targets` is generated after we sort the captions in this batch in decreasing order. That is why I added it here. I can move the `captions` to `model_output`.

vedanuj commented on pythia/modules/attention.py:
```diff
@@ -156,3 +156,25 @@ def forward(self, image_feat, question_embedding, image_locs=None):
         return masked_attention
+
+
+class CaptionAttention(nn.Module):
```
`ImageEmbedding` implementation uses `question_embedding` which we do not require for captioning. Also `AttentionLayer` uses modal combine layer for image and question embedding. We cannot use that as is here. I will look into if we can reuse any piece from there.

apsdehal replied:
But aren't you kind of conditioning on current word? Then question_embedding can be embedding of that word.

vedanuj commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+            data['state'] = {"td_hidden": (h1, c1), "lm_hidden": (h2, c2)}
+        return data
+
+    def beam_search(self, sample_list, k=5):
```
Will move this to utils. Currently `k` is not configurable from anywhere. We can add that as a config option.

vedanuj commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+    def build(self):
+        self._build_word_embedding()
+        self._init_feature_encoders("image")
+        self._init_decoder()
```
I will try to modularize this and move the `TopDownAttention` part outside the decoder and make the decoder a configurable from configs. I will add these changes after I fix the other ones as this will require retraining the models and checking if we maintain the accuracy after the refactor.

vedanuj commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+        sample_list.add_field("targets", targets)
+        model_output = {"scores": scores}
+        del (data)
+        gc.collect()
```
I will remove this. Not necessary actually.

vedanuj commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+
+
+@registry.register_model("butd")
+class BUTD(BaseModel):
```
Sounds good. Refactoring this will need retraining the current model we have. I will add these changes after I finish training with changes in this part.

abhiskk commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+from pythia.modules.decoders import CaptioningDecoder
+from torch import nn
+
+from .base_model import BaseModel
```
Is it possible to avoid relative import here: `from pythia.models.base_model import BaseModel`

abhiskk commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+            {"params": self.decoder.parameters()},
+            {
+                "params": self.image_feature_encoders.parameters(),
+                "lr": (config["optimizer_attributes"]["params"]["lr"] * 0.1),
```
I might be missing a detail here but why is the `lr` being multiplied by 0.1 here?

apsdehal replied:
This layer comes from faster-rcnn and is being finetuned. That's why LR for fine-tuning is 0.1 times lower than the normal LR.

abhiskk replied:
Maybe move this value also to `config`, 0.1 here looks like a magic number.

abhiskk commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+        data = {}
+        data["texts"] = torch.LongTensor([[self.vocab.SOS_INDEX]] * k).to("cuda")
+        data["image_features"] = (
+            image_embedding_total.unsqueeze(1).expand(-1, k, -1, -1).squeeze(0)
```
A comment here would be helpful

abhiskk commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+                )
+
+            # Convert to vocab indices
+            prev_word_inds = top_k_words / self.vocab_size
```
This will return a float, unless `top_k_words` is always divisible by `self.vocab_size`. Suggestion: `prev_word_inds = top_k_words // self.vocab_size`

abhiskk commented on pythia/models/butd.py:
```diff
@@ -0,0 +1,294 @@
+
+
+        if len(complete_seqs_scores) == 0:
+            captions = torch.FloatTensor([0] * timesteps).unsqueeze(0)
```
Use `torch.zeros` here.

abhiskk commented on pythia/modules/attention.py:
```diff
@@ -156,3 +156,25 @@ def forward(self, image_feat, question_embedding, image_locs=None):
+
+class CaptionAttention(nn.Module):
+    def __init__(self, features_dim, decoder_dim, attention_dim, dropout=0.5):
+        super(CaptionAttention, self).__init__()
```
nit: `super().__init__()`

abhiskk commented on pythia/modules/attention.py:
```diff
@@ -156,3 +156,25 @@ def forward(self, image_feat, question_embedding, image_locs=None):
+        self.softmax = nn.Softmax(dim=1)
+
+    def forward(self, image_features, decoder_hidden):
+        att1 = self.features_att(image_features)
```
NIT: `att` -> `attn` is a better usage I think.

abhiskk commented on pythia/modules/attention.py:
```diff
@@ -156,3 +156,25 @@ def forward(self, image_feat, question_embedding, image_locs=None):
+        alpha = self.softmax(att)
+        attention_weighted_encoding = (image_features * alpha.unsqueeze(2)).sum(dim=1)
+
+        return attention_weighted_encoding
```
NIT: be consistent in usage of `attn`, either stick to `attn` everywhere or `attention`

abhiskk commented on pythia/modules/decoders.py:
```diff
@@ -2,6 +2,8 @@
 import torch
 from torch import nn
 
+from .attention import CaptionAttention
```
If possible use absolute import

abhiskk commented on pythia/tasks/captioning/coco/__init__.py:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
The header message seems incomplete here

vedanuj replied:
Different files have different headers. Which is the correct one we should use?

apsdehal replied:
This one works. It is short form of the original header used everywhere.

abhiskk commented on pythia/tasks/captioning/coco/dataset.py:
```diff
@@ -0,0 +1,152 @@
+            return paths
+        else:
+            raise TypeError(
+                "Paths passed to dataset should either be " "string or list"
```
Strong continuation looks odd here, I think this is a `black` artifact

abhiskk commented on pythia/tasks/captioning/coco/dataset.py:
```diff
@@ -0,0 +1,152 @@
+            reference_list.append(processed_reference["text"])
+
+        # Restrict to 5 reference captions
+        sample.answers = torch.stack(reference_list)[:5]
```
Make this a parameter instead of a constant

apsdehal commented on requirements.txt:
```diff
@@ -10,3 +10,4 @@ PyYAML>=3.11
 pytest==3.3.2
 requests==2.21.0
 fastText
+nltk
```
Let's add the version number of nltk here.

vedanuj commented:
@apsdehal @abhiskk I have addressed most of the comments and restructured the code to reuse existing code as much as possible and I believe this is in a better state now. Would like to have your review on this. There are a few minor comments which I still have to address. Will do it soon.

vedanuj commented:
Updated the PR with all requested revisions.

apsdehal commented on configs/captioning/coco/butd.yml:
```diff
@@ -0,0 +1,64 @@
+        bias_file: detectron/fc6/fc7_b.pkl
+        weights_file: detectron/fc6/fc7_w.pkl
+    inference:
+      type: greedy
```
Should we make it beam_search by default? If not, we should add some documentation for this

vedanuj replied:
Currently the beam search works only with batch size 1. So adding it as the default inference type is not feasible during training for now. I will create a separate config file that can run beam search inference along with some comments.

apsdehal commented on pythia/modules/layers.py:
```diff
@@ -108,6 +111,37 @@ def forward(self, x):
         return logits
 
 
+class LanguageDecoder(nn.Module):
```
Maybe move this to `decoders.py` as it is a decode.

vedanuj replied:
I added it here since this is a `ClassifierLayer` type module and other classifier layers are also in this file.

apsdehal commented on pythia/modules/layers.py:
```diff
@@ -273,6 +309,45 @@ def forward(self, image_feat, question_embedding, context_embedding=None):
         return joint_feature
 
 
+class TopDownAttentionLSTM(nn.Module):
```
This should be moved to `attention.py`

vedanuj replied:
I have added `TopDownAttentionLSTM` as a type of `ModalCombineLayer` so that we can reuse the `TopDownAttention` module when we process the feature and word embeddings. Does it still make sense to move to `attention.py` as this is only a type of layer?

apsdehal replied:
Ok, keeping this here makes sense.

apsdehal commented on pythia/modules/metrics.py:
```diff
@@ -202,6 +202,59 @@ def calculate(self, sample_list, model_output, *args, **kwargs):
 
 
+@registry.register_metric("caption_bleu4")
+class CaptionBleu4Metric(BaseMetric):
```
Are we providing support for other metrics? SPICE and ROGUE?

vedanuj replied:
I have requested a separate PR #78 which adds a script for calculating metrics. If we are ok with importing the dependent libraries for calculating these metrics then we can add support for all of them in `metrics.py`.

apsdehal commented on pythia/utils/beam_search.py:
```diff
@@ -0,0 +1,111 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
As discussed, let's move this to text_utils.

vedanuj commented:
Thanks for the review. Should we add the tests in this PR or a separate PR?

apsdehal commented:
Separate is fine. Let's finish this one's comments and merge it. You should also make README, documentation and other changes in a separate PR.

apsdehal commented:
This has landed into master 🎉. Pending TODOs:
- Update README and documentation updates #79 
- Add tests for captioning #80
